File: sdk/go/audio.go
package rtvbp

import (
	"errors"
	"fmt"
	"io"
	"runtime"
	"sync"

	"github.com/smallnest/ringbuffer"
)

// audioStream is the session-owned byte view over one duplex media channel.
// Incoming media is written to inbound; handler writes are read from outbound by the media pump.
type audioStream struct {
	inbound  *ringbuffer.RingBuffer
	outbound *ringbuffer.RingBuffer

	formatMu  sync.RWMutex
	format    MediaFormat
	formatSet bool
	closeOnce sync.Once
}

func newAudioStream(bufferSize int) *audioStream {
	if bufferSize <= 0 {
		panic("rtvbp: audio buffer size must be positive")
	}
	return &audioStream{
		inbound:  ringbuffer.New(bufferSize).SetBlocking(true),
		outbound: ringbuffer.New(bufferSize).SetBlocking(true),
	}
}

func (s *audioStream) Read(p []byte) (int, error) {
	return s.inbound.Read(p)
}

func (s *audioStream) Write(p []byte) (int, error) {
	return s.outbound.Write(p)
}

func (s *audioStream) ClearReadBuffer() (int, error) {
	return discardAvailable(s.inbound)
}

func (s *audioStream) Format() MediaFormat {
	s.formatMu.RLock()
	defer s.formatMu.RUnlock()
	return s.format
}

// setFormat records the immutable format selected by media negotiation. Repeating the same
// selection is harmless, while a different second selection is rejected.
func (s *audioStream) setFormat(format MediaFormat) error {
	if _, err := format.FrameBytes(); err != nil {
		return err
	}
	s.formatMu.Lock()
	defer s.formatMu.Unlock()
	if s.formatSet {
		if s.format == format {
			return nil
		}
		return fmt.Errorf("rtvbp: audio format is already negotiated as %#v", s.format)
	}
	s.format = format
	s.formatSet = true
	return nil
}

func (s *audioStream) frameBytes() (int, error) {
	s.formatMu.RLock()
	defer s.formatMu.RUnlock()
	if !s.formatSet {
		return 0, fmt.Errorf("rtvbp: audio format is not negotiated")
	}
	return s.format.FrameBytes()
}

func (s *audioStream) writeInbound(p []byte) (int, error) {
	return s.inbound.Write(p)
}

func (s *audioStream) readOutbound(p []byte) (int, error) {
	return s.outbound.Read(p)
}

func (s *audioStream) Close() error {
	s.closeOnce.Do(func() {
		s.inbound.CloseWithError(io.EOF)
		s.outbound.CloseWithError(io.EOF)
	})
	return nil
}

func discardAvailable(buffer *ringbuffer.RingBuffer) (int, error) {
	remaining := buffer.Length()
	if remaining == 0 {
		return 0, nil
	}
	scratch := make([]byte, min(remaining, 32*1024))
	cleared := 0
	for cleared < remaining {
		want := min(len(scratch), remaining-cleared)
		n, err := buffer.TryRead(scratch[:want])
		cleared += n
		switch {
		case err == nil:
		case errors.Is(err, ringbuffer.ErrAcquireLock):
			runtime.Gosched()
		case errors.Is(err, ringbuffer.ErrIsEmpty):
			return cleared, nil
		default:
			return cleared, err
		}
	}
	return cleared, nil
}

// AudioChannelSide is a compatibility view over the session-owned audio stream.
// New session/media code should own audioStream directly.
type AudioChannelSide struct {
	stream  *audioStream
	reverse bool
}

// Close closes the reader and writer
func (s *AudioChannelSide) Close() error {
	return s.stream.Close()
}

func (s *AudioChannelSide) Read(p []byte) (n int, err error) {
	if s.reverse {
		return s.stream.readOutbound(p)
	}
	return s.stream.Read(p)
}

func (s *AudioChannelSide) Write(p []byte) (n int, err error) {
	if s.reverse {
		return s.stream.writeInbound(p)
	}
	return s.stream.Write(p)
}

func (s *AudioChannelSide) ClearReadBuffer() (int, error) {
	if s.reverse {
		return discardAvailable(s.stream.outbound)
	}
	return s.stream.ClearReadBuffer()
}

func (s *AudioChannelSide) Format() MediaFormat {
	return s.stream.Format()
}

func NewAudioChannel(audioBufferSize int) (*AudioChannelSide, *AudioChannelSide) {
	stream := newAudioStream(audioBufferSize)
	return &AudioChannelSide{stream: stream}, &AudioChannelSide{stream: stream, reverse: true}
}

File: sdk/typescript/src/audio.ts
import { AsyncQueue } from "./async.ts";
import { SessionError, asSessionError, throwIfAborted } from "./errors.ts";
import type { MediaChannel, MediaFormat, MediaFrame } from "./transport.ts";
import { mediaFrameBytes } from "./transport.ts";

export interface ObservedMediaFrame {
  readonly direction: "in" | "out";
  readonly data: Uint8Array;
  readonly ptsMs?: number;
  readonly observedAt: number;
}

export type AudioObserver = (frame: ObservedMediaFrame) => void;

function sameFormat(left: MediaFormat, right: MediaFormat): boolean {
  return left.encoding === right.encoding
    && left.sampleRate === right.sampleRate
    && left.bitDepth === right.bitDepth
    && left.channels === right.channels
    && left.packetTimeMs === right.packetTimeMs;
}

/** Session-owned byte stream over one negotiated, fixed-size media channel. */
export class AudioStream {
  readonly #capacityBytes: number;
  readonly #observer: AudioObserver | undefined;
  readonly #onError: ((error: SessionError) => void) | undefined;
  readonly #abort = new AbortController();
  #channel: MediaChannel | undefined;
  #format: MediaFormat | undefined;
  #inbound: AsyncQueue<MediaFrame> | undefined;
  #current: Uint8Array | undefined;
  #currentOffset = 0;
  #partial = new Uint8Array(0);
  #writeTail: Promise<void> = Promise.resolve();
  #readTail: Promise<void> = Promise.resolve();
  #reader: Promise<void> | undefined;
  #closed = false;

  constructor(
    capacityBytes = 1024 * 1024,
    observer?: AudioObserver,
    onError?: (error: SessionError) => void,
  ) {
    if (!Number.isSafeInteger(capacityBytes) || capacityBytes <= 0) {
      throw new SessionError("configuration", "audio buffer capacity must be positive");
    }
    this.#capacityBytes = capacityBytes;
    this.#observer = observer;
    this.#onError = onError;
  }

  get format(): MediaFormat | undefined {
    return this.#format;
  }

  get bufferedInboundBytes(): number {
    const current = this.#current === undefined ? 0 : this.#current.byteLength - this.#currentOffset;
    if (this.#format === undefined || this.#inbound === undefined) return current;
    return current + this.#inbound.length * mediaFrameBytes(this.#format);
  }

  /** Bind exactly once. Repeating the same channel is harmless; format changes fail closed. */
  bind(channel: MediaChannel): void {
    if (this.#closed) throw new SessionError("closed", "audio stream is closed");
    mediaFrameBytes(channel.format);
    if (this.#format !== undefined && !sameFormat(this.#format, channel.format)) {
      throw new SessionError("media_format", "audio format is already negotiated");
    }
    if (this.#channel !== undefined) {
      if (this.#channel !== channel) throw new SessionError("media_duplicate", "audio is already bound");
      return;
    }
    this.#channel = channel;
    this.#format = { ...channel.format };
    const frames = Math.max(1, Math.floor(this.#capacityBytes / mediaFrameBytes(channel.format)));
    this.#inbound = new AsyncQueue<MediaFrame>(frames);
    this.#reader = this.#readFrames(channel, this.#inbound).catch((error: unknown) => {
      const failure = asSessionError(error, "media_read");
      this.#onError?.(failure);
      throw failure;
    });
    void this.#reader.catch(() => {});
  }

  async write(data: Uint8Array, signal?: AbortSignal): Promise<number> {
    throwIfAborted(signal);
    if (data.byteLength === 0) return 0;
    const copy = data.slice();
    const previous = this.#writeTail;
    let release: () => void = () => {};
    this.#writeTail = new Promise<void>((resolve) => { release = resolve; });
    await previous;
    try {
      throwIfAborted(signal);
      const channel = this.#requireChannel();
      const frameBytes = mediaFrameBytes(channel.format);
      const combined = new Uint8Array(this.#partial.byteLength + copy.byteLength);
      combined.set(this.#partial);
      combined.set(copy, this.#partial.byteLength);
      let offset = 0;
      while (combined.byteLength - offset >= frameBytes) {
        throwIfAborted(signal);
        const frame = combined.slice(offset, offset + frameBytes);
        const ptsMs = Date.now();
        await channel.writeFrame({ data: frame, ptsMs }, signal);
        this.#observe({ direction: "out", data: frame, ptsMs, observedAt: Date.now() });
        offset += frameBytes;
      }
      this.#partial = combined.slice(offset);
      return copy.byteLength;
    } finally {
      release();
    }
  }

  async read(maxBytes: number, signal?: AbortSignal): Promise<Uint8Array> {
    if (!Number.isSafeInteger(maxBytes) || maxBytes <= 0) {
      throw new SessionError("configuration", "read size must be positive");
    }
    throwIfAborted(signal);
    const previous = this.#readTail;
    let release: () => void = () => {};
    this.#readTail = new Promise<void>((resolve) => { release = resolve; });
    await previous;
    try {
      throwIfAborted(signal);
      const inbound = this.#requireInbound();
      if (this.#current === undefined || this.#currentOffset === this.#current.byteLength) {
        this.#current = (await inbound.shift(signal)).data;
        this.#currentOffset = 0;
      }
      const end = Math.min(this.#currentOffset + maxBytes, this.#current.byteLength);
      const result = this.#current.slice(this.#currentOffset, end);
      this.#currentOffset = end;
      return result;
    } finally {
      release();
    }
  }

  clear(): number {
    const cleared = this.bufferedInboundBytes;
    this.#current = undefined;
    this.#currentOffset = 0;
    this.#inbound?.clear();
    return cleared;
  }

  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    this.#abort.abort(new SessionError("closed", "audio stream is closed"));
    this.#inbound?.close();
    await this.#channel?.close();
    await this.#reader?.catch(() => {});
    this.#partial = new Uint8Array(0);
  }

  async #readFrames(channel: MediaChannel, inbound: AsyncQueue<MediaFrame>): Promise<void> {
    try {
      while (!this.#abort.signal.aborted) {
        const frame = await channel.readFrame(this.#abort.signal);
        if (frame.data.byteLength !== mediaFrameBytes(channel.format)) {
          throw new SessionError("media_frame", "received media frame has the wrong byte length");
        }
        const copy = {
          data: frame.data.slice(),
          ...(frame.ptsMs === undefined ? {} : { ptsMs: frame.ptsMs }),
        };
        this.#observe({
          direction: "in",
          data: copy.data,
          ...(copy.ptsMs === undefined ? {} : { ptsMs: copy.ptsMs }),
          observedAt: Date.now(),
        });
        await inbound.push(copy, this.#abort.signal);
      }
    } catch (error) {
      if (!this.#abort.signal.aborted) {
        inbound.close();
        throw asSessionError(error, "media_read");
      }
    }
  }

  #observe(frame: ObservedMediaFrame): void {
    this.#observer?.({ ...frame, data: frame.data.slice() });
  }

  #requireChannel(): MediaChannel {
    if (this.#closed) throw new SessionError("closed", "audio stream is closed");
    if (this.#channel === undefined) throw new SessionError("media_unbound", "audio is not negotiated");
    return this.#channel;
  }

  #requireInbound(): AsyncQueue<MediaFrame> {
    this.#requireChannel();
    if (this.#inbound === undefined) throw new SessionError("media_unbound", "audio is not negotiated");
    return this.#inbound;
  }
}
